Subject: 2 timestamp fields in TestExecution class #1924

APiligrim opened the pull request:
This commit is part of the issue #1924 where Kiwi TCMS needs additional fields for the TestExecution class. This is a prerequisite for time tracking features & telemetry. Here are the changes that this commit is trying to make:

- New field TestExecution.start_date of type DateTimeField(db_index=True, null=True, blank=True)

- Renamed the TestExecution.close_date to TestExecution.stop_date and add db_index=True

- Added a new property TestExecution.actual_duration - the difference between the 2 fields

- Added a new property TestExecution.expected_duration -> self.case.expected_duration; for convenience, see #1923

- API method TestExecution.update() supports the new start_date and close_ date fields

atodorov reviewed (changes requested):
* Run `manage.py makemigrations` & squash all commits together, see http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html
* See https://chris.beams.io/posts/git-commit/ on info how to improve your commit log message
* See https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords about how to reference existing issues in commit log

APiligrim commented:
Im having some trouble😅 for some reason although I pulled from master, then ran manage.py makemigrations, made another commit and squashed it with all the previous ones I made, calling it "Add: New timestamp fields in TestExecution class .Refs #1924", its still trying to commit them twice. Also the commit with "Merge branch 'anastasiya-branch' of https://github.com/MLH-Fellowship… …" is not showing up when I search for it with rebase log. Do you have any advice on how I can delete the repeated commits that are showing up here? Thank you for all your help!

atodorov commented on tcms/testruns/migrations/0012_auto_20210217_0446.py:
```diff
@@ -0,0 +1,41 @@
+# Generated by Django 3.1.6 on 2021-02-17 04:46
```
Change the auto-generated file name to 0012_<something_more_meaningful>.py

atodorov commented on tcms/testruns/migrations/0012_auto_20210217_0446.py:
```diff
@@ -0,0 +1,41 @@
+        migrations.RemoveField(
+            model_name='testexecution',
+            name='close_date',
+        ),
```
This will break existing installations and destroy information. See `migrations.RenameField`

atodorov commented:
> Im having some troublesweat_smile for some reason although I pulled from master, then ran manage.py makemigrations, made another commit and squashed it with all the previous ones I made, calling it "Add: New timestamp fields in TestExecution class .Refs #1924", its still trying to commit them twice. Also the commit with "Merge branch 'anastasiya-branch' of https://github.com/MLH-Fellowship… …" is not showing up when I search for it with rebase log. Do you have any advice on how I can delete the repeated commits that are showing up here? Thank you for all your help!

The commit you see with "Merge branch ..." is called a merge commit, aka 3 way merge (https://hamwaves.com/collaboration/doc/rypress.com/branches-2.html search for 3-way merge).

This happened automatically when you did `git pull` b/c that means download the new commits and automatically merge them into the current branch. The workflow is usually a bit different:
```
git checkout master
git fetch upstream  <--- assumes you have another remote configured 
git merge upstream/master  <---- keeps the local master branch clean, performs a fast-forward merge without merge commits
git checkout <your_topic_branch>
git rebase -i master  <--- rebases onto the refreshed local master
```

Another option if the local master branch is not clean is to rebase against `upstream/master` after a fetch! (There are others as well). 


Don't sweat it too much. The upstream branch is moving very rapidly but unless you need to rebase because of conflicts you can keep working against an older version as well.

atodorov commented on tcms/testruns/migrations/0012_test_execusion_add_fields.py:
```diff
@@ -0,0 +1,43 @@
+            model_name="historicaltestexecution",
+            name="stop_date",
+            field=models.DateTimeField(blank=True, db_index=True, null=True),
+        ),
```
This is causing 
`django.db.utils.ProgrammingError: column "stop_date" of relation "testruns_historicaltestexecution" already exists` in test-migrations-rollback. Remove it.

APiligrim replied:
I checked **test-migrations-rollback**  but I cant find anywhere where it `references the stop_date of relation "testruns_historicaltestexecution" already exists ` so that I can remove it. 

This is what that it shows: 
` test-migrations-rollback:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: [3.7]

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@v1
      with:
        python-version: ${{ matrix.python-version }}

    - name: Create database
      run: |
        docker-compose -f docker-compose.postgres pull db
        docker-compose -f docker-compose.postgres run -d -p 5432:5432 --name kiwi_db db

    - name: migrations rollback
      run: |
        pip install -r requirements/devel.txt
        pip install -r requirements/postgres.txt
        pushd tcms/ && npm install && popd

        docker exec -i kiwi_db psql -c "ALTER USER kiwi CREATEDB;"

        export LANG=en-us
        # Postgres b/c it has best support for schema altering operations
        export DJANGO_SETTINGS_MODULE=tcms.settings.test.postgresql
        ./tests/migrations-rollback`

Do you know of any other ways to check where this duplicated column is created? I also searched where else the stop_date is being referenced but I'm not sure which specific code creates that extra column.

atodorov replied:
Migration operations are executed in the order they are defined in the migration file. You are renaming an older field to `stop_date` 15-20 lines above and as a last operation you are trying to add a new field with the same name which obviously doesn't work. 

This is a left-over from your 1st and 2nd commits. Checkout your git log/individual commits for history.

atodorov commented on tcms/testruns/models.py:
```diff
@@ -256,9 +257,21 @@ class TestExecution(TCMSActionModel):
+        duration_difference = self.start_date - self.stop_date
+        return duration_difference
+
+    actual_duration = property(_actual_duration)
```
I prefer the 
```
@property
def actual_duration(self):
``

syntax for properties.

atodorov commented on tcms/testruns/models.py:
```diff
@@ -143,7 +143,8 @@ def create_execution(
-            close_date=None,
+            stop_date=None,
+            start_date=None,
         )
```
you are changing the signature method here which means you have to update all invocation of this method as well.

atodorov commented on tcms/testruns/models.py:
```diff
@@ -256,9 +257,21 @@ class TestExecution(TCMSActionModel):
+    def _expected_duration(self):
+        return self.case.expected_duration
+
+    expected_duration = property(_expected_duration)
```
This is blocked on #2253 b/c at the moment `case.expected_duration` does not exist.

atodorov commented on tcms/testruns/static/testruns/js/get.js:
```diff
@@ -383,7 +383,7 @@ function bindEvents(selector) {
     const container = $(`.test-execution-${testExecution.id}`)
 
-    container.find('.test-execution-information .run-date').html(testExecution.close_date || '-')
+    container.find('.test-execution-information .run-date').html(testExecution.stop_date || '-')
```
This will very likely fail b/c of other conflicting changes in the API. It's best if you focus on the backend changes first and leave this for later. We may need to push this change to a different branch.

atodorov commented on tcms/rpc/api/testexecution.py:
```diff
@@ -185,9 +185,11 @@ def update(execution_id, values, **kwargs):
         if test_execution.status.weight != 0:
-            test_execution.close_date = now
+            test_execution.stop_date = now
+            test_execution.start_date = now
```
Don't modify `start_date` here nor below!

I'm not sure if we need to do this ATM.

atodorov commented on tcms/rpc/api/testexecution.py:
```diff
@@ -185,9 +185,11 @@ def update(execution_id, values, **kwargs):
         now = timezone.now()
         if test_execution.status.weight != 0:
-            test_execution.close_date = now
+            test_execution.stop_date = now
```
Changes for stop_date will need testing, but let's focus on the models first!

atodorov commented on tcms/testruns/models.py:
```diff
@@ -256,9 +257,21 @@ class TestExecution(TCMSActionModel):
     case_text_version = models.IntegerField()
-    close_date = models.DateTimeField(null=True, blank=True)
+    start_date = models.DateTimeField(null=True, blank=True, db_index=True)
+    stop_date = models.DateTimeField(null=True, blank=True, db_index=True)
```
These fields also require TestExecutionFactory to be updated otherwise they cause the tests to fail.

APiligrim commented:
Hi @atodorov! I renamed all the associated "close_date" vars to "stop_date" and I also removed the "case.expected_duration " which refers to  #2253 and is b/c at the moment. I see more checks are passing but some are still failing so I'll explore some of those issues 👀

atodorov reviewed (changes requested):
Drop the last commit and focus on the other comments from the code review. 

Renaming `close_date` to `stop_date` should be done where needed, not everywhere where this variable name occurs. As-is the last commit is creating more issues than it solves.

atodorov commented on tcms/rpc/api/testexecution.py:
```diff
@@ -196,9 +196,11 @@ def update(execution_id, values, **kwargs):
             and not all_executions.filter(status__weight=0).exists()
         ):
             test_execution.run.stop_date = now
+            test_execution.run.start_date = now
```
Remove this line, we don't want to reset `start_date` after a test execution has been created.

atodorov commented on tcms/rpc/api/testexecution.py:
```diff
@@ -196,9 +196,11 @@ def update(execution_id, values, **kwargs):
             test_execution.run.save()
         elif test_execution.status.weight == 0 and test_execution.run.stop_date:
             test_execution.run.stop_date = None
+            test_execution.run.start_date = None
```
Remove this line, we don't want to reset `start_date` after a test execution has been created.

atodorov commented on tcms/rpc/serializer.py:
```diff
@@ -428,7 +428,7 @@ class TestExecutionRPCSerializer(QuerySetBasedRPCSerializer):
         "id": ("id", do_nothing),
         "case_text_version": ("case_text_version", do_nothing),
-        "close_date": ("close_date", datetime_to_str),
+        "stop_date": ("stop_date", datetime_to_str),
```
Remove this. The entire serializer module has been removed in the wip_remove_serializer branch.

atodorov commented on tcms/testruns/migrations/0004_squashed.py:
```diff
@@ -60,7 +60,7 @@ class Migration(migrations.Migration):
                 ("case_text_version", models.IntegerField()),
                 ("running_date", models.DateTimeField(blank=True, null=True)),
-                ("close_date", models.DateTimeField(blank=True, null=True)),
+                ("stop_date", models.DateTimeField(blank=True, null=True)),
```
Undo the changes to this file. Historical migrations generally do not need to be modified.

You can read more about migrations at https://docs.djangoproject.com/en/3.1/topics/migrations/

atodorov replied:
Still unresolved.

APiligrim replied:
I changed the names of "stop_date" back to "close_date" as they were written before. I'm not really sure what else I should change since thats the only thing that was changed in this historic file 👀

atodorov replied:
> I manually changed the names of "stop_date" back to "close_date" as they were written before. I'm not really sure what else I should change since thats the only thing that was changed in this historic file

Well if that's the only thing which changed and I tell you that it is not supposed to be there then when you update the file for a second time it should look exactly like it was before.

> Do I have to undo the change to this file using the git commands?

I don't follow what you mean but in any case all changes made locally should be committed and pushed to GitHub.

atodorov commented on tcms/testruns/migrations/0012_test_execusion_add_fields.py:
```diff
@@ -0,0 +1,38 @@
+    operations = [
+        migrations.RenameField(
+            model_name="historicaltestexecution",
+            old_name="stop_date",
```
Old name was close_date, wasn't it ?

atodorov commented on tcms/testruns/migrations/0012_test_execusion_add_fields.py:
```diff
@@ -0,0 +1,38 @@
+        ),
+        migrations.RenameField(
+            model_name="testexecution",
+            old_name="stop_date",
```
Old name was close_date, wasn't it ?

atodorov commented on tcms/rpc/tests/test_testexecution.py:
```diff
@@ -676,10 +676,10 @@ def test_when_build_specified_does_not_change_build(self):
 
     def test_non_zero_status_changes_close_date(self):
```
Update method name to reflect new field name.

atodorov commented on tcms/rpc/tests/test_testexecution.py:
```diff
@@ -690,14 +690,14 @@ def test_non_zero_status_changes_close_date(self):
-        self.assertGreater(self.execution_1.close_date, few_secs_ago)
+        self.assertGreater(self.execution_1.stop_date, few_secs_ago)
 
     def test_zero_status_changes_close_date(self):
```
Update method name to reflect new field name.

atodorov commented on tcms/testruns/models.py:
```diff
@@ -256,9 +257,15 @@ class TestExecution(TCMSActionModel):
     sortkey = models.IntegerField(null=True, blank=True)
 
+    @property
+    def _actual_duration(self):
```
Name should not start with `_`

atodorov commented on tcms/testruns/models.py:
```diff
@@ -256,9 +257,15 @@ class TestExecution(TCMSActionModel):
 
+    @property
+    def _actual_duration(self):
+        duration_difference = self.start_date - self.stop_date
```
You don't need an intermediate variable to store the results, `return` the expression directly.

atodorov replied:
stop_date > start_date, adjust the expression.

atodorov commented on tcms/testruns/migrations/0012_test_execusion_add_fields.py:
```diff
@@ -0,0 +1,33 @@
+            model_name="testexecution",
+            old_name="close_date",
+            new_name="stop_date",
+        ),
```
The renamed field also has `db_index=True`, see models.py which means the migration also needs an AlterField operation on these fields to add the index.

APiligrim replied:
When I added the additional migrations:
`
        migrations.AlterField(
            model_name="testexecution",
            name="stop_date",
            field=models.AutoField(primary_key=True, serialize=False),
        ),
        migrations.AlterField(
            model_name="testexecution",
            name="start_date",
            field=models.AutoField(primary_key=True, serialize=False),
        ),
`
there pops up an error that "AssertionError: Model testruns.TestExecution can't have more than one auto-generated field." do you have any advice as to where I should look to make it work?

atodorov replied:
The error is self explanatory, you can't have 2 primary keys in a table. 

Before going further you should think about what you are trying to do and not try to guess it because I see that it isn't very clear to you what you are doing. 

1) My last comment talks about only the renamed field, which is now called `stop_date`. Your last example tries to modify 2 fields instead of one !!! Pay attention to this!

2) The immediate goal for `stop_date` field after it was renamed is to instruct the database engine to create an index for that field. Index has been missing up until now. IDK how you came up with the idea that modifying the field to be a primary key will alter its index but it's not what you want to do. Explore the other migrations in the source code, there are   at least 2 examples of the same operation.  Also read the documentation about AlterField -> https://docs.djangoproject.com/en/3.1/ref/migration-operations/#alterfield

atodorov reviewed (changes requested):
See latest comments & squash all commits together.

atodorov commented on tcms/testruns/migrations/0004_squashed.py:
```diff
@@ -480,4 +480,4 @@ class Migration(migrations.Migration):
             },
         ),
-    ]
+    ]
```
Alright, I finally figured out what's going on here. Line 114 is `TestRun.stop_date`. That field is unrelated to the current issue and it has the appropriate name from the start. No changes should be made there.

Line 397 is similar but for HistoricalTestRun. 

@APiligrim  I've made a confusing comment before but you have overlooked the fact that these migration operations are towards a completely different model.

atodorov replied:
Revert all changes to this file, it should not show up in the diff at all.

APiligrim replied:
Thank you! I reverted my changes to that specific file back to be like what is on master 👍

atodorov commented on tcms/testruns/migrations/0012_test_execusion_add_fields.py:
```diff
@@ -0,0 +1,43 @@
+        migrations.AlterField(
+            model_name="testexecution",
+            name="stop_date",
+            field=models.DateTimeField(db_index=True, null=True),
```
Causes `sanity / test-for-missing-migrations` to fail b/c the way you alter the fields in the DB results in them being slightly different from the models. See what the failing test does and replicate it by hand to see the difference.

atodorov reviewed (changes requested):
Minor changes requested!

- Squash commits and update commit log to match!

atodorov reviewed (approved):
LGTM, docker test failed for unrelated reasons. 

As soon as it passes I will merge this PR, check the issue for the next steps.
